map_gui/src/simple_app.rs

```diff
@@ -287,3 +287,39 @@ impl State<SimpleApp> for SimpleWarper {
 
     fn draw(&self, _: &mut GfxCtx, _: &SimpleApp) {}
 }
+
+/// Store a cached key/value pair, only recalculating when the key changes.
+pub struct Cached<K, V> {
+    contents: Option<(K, V)>,
+}
+
+impl<K: PartialEq + Clone, V> Cached<K, V> {
```

Comment:
Will we provide nicer compiler errors if we put these constraints on the struct as well?

Reply:
Done. I've been bitten by that before for sure. I guess it's unlikely somebody would pass around a `Cached<T>` in a different module than where they call `update`, but having the error show up immediately seems nicer.